include/pybind11/pybind11.h

```diff
@@ -1986,43 +2010,69 @@ iterator make_iterator(Iterator first, Sentinel last, Extra &&... extra) {
                     s.first_or_done = true;
                     throw stop_iteration();
                 }
-                return *s.it;
+                return Access()(s.it);
             // NOLINTNEXTLINE(readability-const-return-type) // PR #3263
             }, std::forward<Extra>(extra)..., Policy);
     }
 
     return cast(state{first, last, true});
 }
 
-/// Makes an python iterator over the keys (`.first`) of a iterator over pairs from a
+PYBIND11_NAMESPACE_END(detail)
+
+/// Makes a python iterator from a first and past-the-end C++ InputIterator.
+template <return_value_policy Policy = return_value_policy::reference_internal,
+          typename Iterator,
+          typename Sentinel,
+#ifndef DOXYGEN_SHOULD_SKIP_THIS  // Issue in breathe 4.26.1
+          typename ValueType = typename detail::iterator_access<Iterator>::result_type,
+#endif
+          typename... Extra>
+iterator make_iterator(Iterator first, Sentinel last, Extra &&... extra) {
+    return detail::make_iterator_impl<
+        detail::iterator_access<Iterator>,
+        Policy,
+        Iterator,
+        Sentinel,
+        ValueType,
+        Extra...>(first, last, std::forward<Extra>(extra)...);
+}
+
+/// Makes a python iterator over the keys (`.first`) of a iterator over pairs from a
 /// first and past-the-end InputIterator.
 template <return_value_policy Policy = return_value_policy::reference_internal,
           typename Iterator,
           typename Sentinel,
 #ifndef DOXYGEN_SHOULD_SKIP_THIS  // Issue in breathe 4.26.1
-          typename KeyType = decltype((*std::declval<Iterator>()).first),
+          typename KeyType = typename detail::iterator_key_access<Iterator>::result_type,
 #endif
```

Comment:
@Skylion007 I don't know what this `breathe` issue was, but maybe my change will resolve it? I can try removing the `#ifndef` but I don't know what symptoms to look for.

Reply:
@bmerry From looking at the git blame, the issue is further described here: https://github.com/pybind/pybind11/pull/2828

Reply:
Thanks. If I've followed the chain of causes, I think it's probably no longer an issue (I've run into the same issue with Doxygen myself, where it eats parentheses from decltype expressions), so I've deleted the DOXYGEN_SHOULD_SKIP_THIS guards and we'll see what the CI says.

Reply:
@bmerry Also, we could consider bumping Breathe if needed. It's currently pinned to do that version.